prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -1785,7 +1785,9 @@ layoutWadlerLeijen
       | fits pWidth minNestingLevel availableWidth x = x
       | otherwise = y
       where
-        minNestingLevel = min lineIndent currentColumn
+        minNestingLevel = case y of
+            SLine _ _ -> lineIndent
+            _         -> currentColumn
```

Comment:
We check whether the second alternative (`y`) starts with a line break in order to detect hanging layouts as (partially) explained in https://github.com/quchen/prettyprinter/issues/83#issuecomment-546120600.

Reply:
The loss of laziness in `y` is a bit unfortunate. Since `best` seems to be quite lazy too, I hope it's not too bad though.

Reply:
> The loss of laziness in `y` is a bit unfortunate.

I think this is probably outweighed by the efficiencies gained from the greater `minNestingLevel` when we select `currentColumn`:

* We don't have to check `x` as far as with `lineIndent`.
* We can more often stick with `x` instead of switching to `y`.

Reply:
Also note that `layoutPretty` is completely unaffected since it ignores the `minNestingLevel`:

https://github.com/quchen/prettyprinter/blob/4d51362d7ea8bca0fe9e7ffec2ccd8fcd0065c0d/prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs#L1644-L1647

Reply:
If `y` starts with one of the annotation constructors, we should probably look further for the first "proper" constructor…

Reply:
I went with the »annotations might mean something important even if they contain something unimportant« elsewhere (strip trailing whitespace), so I think ignoring annotations here is alright. What do you think?

Reply:
I'm actually not really familiar with how annotations work or how they are used.

But it seems to me that the result of `layoutWadlerLeijen` should not depend on the presence of annotations. Wouldn't the following property be desirable?

    layoutWadlerLeijen fp opts (unAnnotate doc) === unAnnotateS (layoutWadlerLeijen fp opts doc)

So I think it would be better to check what's behind the annotations.